akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala

```diff
@@ -0,0 +1,127 @@
+    }
+
+    "work when source is exactly the required prefix" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .prefixAndDownstreamMat(10){ prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        } (Keep.right)
+        .toMat(Sink.seq) (Keep.both)
+        .run
+
+      prefixF.futureValue should === (0 until 10)
+      suffixF.futureValue should be (empty)
+    }
+
+    "work when source has less than the required prefix" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .prefixAndDownstreamMat(20){ prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        } (Keep.right)
+        .toMat(Sink.seq) (Keep.both)
+        .run
+
+      prefixF.futureValue should === (0 until 10)
+      suffixF.futureValue/*(Interval(Span(1000, Minutes)))*/ should be (empty)
+    }
+
+    "simple identity case when downstream completes before consuming the entire stream" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source(0 until 100)
+        .prefixAndDownstreamMat(10){ prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        } (Keep.right)
+        .take(10)
+        .toMat(Sink.seq) (Keep.both)
+        .run
+
+      prefixF.futureValue should === (0 until 10)
+      suffixF.futureValue/*(Interval(Span(1000, Minutes)))*/ should === (10 until 20)
+    }
+
+    "propagate materialization failure" in assertAllStagesStopped {
+      val suffixF = Source(0 until 100)
+        .prefixAndDownstreamMat(10){ prefix =>
+          sys error s"I hate mondays! (${prefix.size})"
+        } (Keep.right)
+        .to(Sink.ignore)
+        .run
+
+      val ex = suffixF.failed.futureValue
+      ex.getCause should not be null
+      ex.getCause should have message ("I hate mondays! (10)")
+    }
+
+    "propagate flow failures" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source(0 until 100)
+        .prefixAndDownstreamMat(10){ prefix =>
+          Flow[Int]
+            .mapMaterializedValue(_ => prefix)
+            .map{
+              case 15 => sys error "don't like 15 either!"
+              case n => n
+            }
+        } (Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run
+      prefixF.futureValue should === (0 until 10)
+      val ex = suffixF.failed.futureValue
+      //ex.printStackTrace()
+      ex should have message ("don't like 15 either!")
+    }
+
+    "produce multiple elements per input" in assertAllStagesStopped {
+      val(prefixF, suffixF) =  src10()
+        .prefixAndDownstreamMat(7){ prefix =>
+          Flow[Int]
+            .mapMaterializedValue(_ => prefix)
+            .mapConcat(n => List.fill(n - 6)(n))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should === (0 until 7)
+      suffixF.futureValue should === (7 :: 8 :: 8 :: 9 :: 9 :: 9 :: Nil)
+
+
+    }
+  }
+
+}
```

Comment:
I must have missed this last pass, good that it's here.

I think it would be good to also cover (please also think and see if you can come up with any other corner case scenarios I have missed):

 * upstream completion without any elements
 * both downstream cancellation and failure 
   * without any elements coming downstream
   * as tight as possible with the prefix coming downstream (it will likely be impossible to control the order of events deterministically but all cases should work so still useful)
   * after nested flow has been materialized
 * abrupt termination (stopping the materializer)

Reply:
@johanandren what does 

> as tight as possible with the prefix coming downstream

mean? can you please point me to a test suite taking this approach?

Reply:
As in doing the downstream cancellation immediately after sending the last prefix element (we don't have control over if the prefix element or the cancellation will reach the stage logic first so therefore it won't be deterministic)

Reply:
ok, I think I got it.
BTW, I'm in the process of adding tests using published and subscriber probes. the first case is: n=2, upstream completes immediately after subscriber requests elements (publisher produces no elements), this unveiled a slew of bugs in my code, will push soon.

does it matter if the publisher completes before or after demand? is it deterministic?

Reply:
Waiting for demand would be a good way to know that at least the stage has pulled. If you complete right away you won't know if the stage did pull or not.

Reply:
Some additional potential good things to cover that I could come up with:

 * already mentioned, but, materialization failing
 *  returned flow cancelling but continuing to emit elements, for example `take(0).concat(Source(1 to 4))`
 * returned flow completing downstream but continuing to consume elements, for example  `fromSinkAndSource(Sink.ignore, Source.empty)` 
 * downstream cancelation and failure, could be good using a `fromSinkAndSource(probe, probe)` as nested flow so that you can observe cancellation/failure coming from downstream and then propagate it
 * abrupt termination (stopping the materializer) see for example last test in SeqSinkSpec